packages/sources/coinmetrics-test/src/endpoint/price-ws.ts

```diff
@@ -0,0 +1,170 @@
+  Provider: {
+    WsMessage: WsAssetMetricsMessage
+  }
+}
+
+export const calculateAssetMetricsUrl = async (
+  context: EndpointContext<WsAssetMetricsEndpointTypes>,
+  desiredSubs:
+    | Promise<WsAssetMetricsEndpointTypes['Request']['Params'][]>
+    | WsAssetMetricsEndpointTypes['Request']['Params'][],
+): Promise<string> => {
+  logger.trace(
+    `Calculating URL based on the following desiredSubs array: ${JSON.stringify(desiredSubs)}`,
+  )
+  // The tail of desired subs has data for the URL we want to connect to
+  // Delete all but the tail from desiredSubs (which will make the transport unsubscribe)
+  const internal = await desiredSubs
+  const sub = internal[internal.length - 1]
+  if (!sub) {
+    logger.warn('No subscription found for websocket url generation')
+    return ''
+  }
+  logger.debug(`Generating websocket url for latest desiredSub ${JSON.stringify(sub)}}`)
+  const params = {
+    assets: sub.base.toLowerCase(),
+    metrics: `ReferenceRate${sub.quote.toUpperCase()}`,
+    frequency: '1s',
+    api_key: context.adapterConfig.API_KEY,
+  }
+  logger.debug(`Generating url with: ${JSON.stringify(params)}`)
+  const generated = new URL(
+    `/v4/timeseries-stream/asset-metrics?${querystring.stringify(params)}`,
+    context.adapterConfig.WS_API_ENDPOINT || DEFAULT_WS_API_ENDPOINT,
+  ).toString()
+  logger.debug(`Generated the following url: ${generated}`)
+  return generated
+}
+
+export const handleAssetMetricsMessage = (
+  message: WsAssetMetricsMessage,
+): ProviderResult<WsAssetMetricsEndpointTypes>[] | undefined => {
+  logger.trace(message, 'Got response from websocket')
+  //Cast message to all of its known types so we can check which variant we received
+  const reorgMessage = message as WsAssetMetricsReorgResponse
+  const warningMessage = message as WsAssetMetricsWarningResponse
+  const successMessage = message as WsAssetMetricsSuccessResponse
+  const errorMessage = message as WsAssetMetricsErrorResponse
+  if (successMessage.asset && successMessage.type !== 'reorg') {
+    // Success message
+    let metric: number | undefined
+    let metricQuote = ''
+    for (const quote of VALID_REFERENCE_RATE_QUOTES) {
+      metric = successMessage[`ReferenceRate${quote}`]
+      if (metric) {
+        metricQuote = quote
+        break
+      }
+    }
+    if (!metric) {
+      //TODO should we throw an error here?
+      logger.warn(
+        successMessage,
+        `Could not find any defined ReferenceRate(${VALID_REFERENCE_RATE_QUOTES.join(
+          '|',
+        )}) properties in message (${JSON.stringify(message)}), will return 0 value as default`,
+      )
+    }
+    return [
+      {
+        params: {
+          base: successMessage.asset,
+          quote: metricQuote,
+          endpoint: 'price-ws',
+        },
+        value: metric || 0, // Fallback zero here just to get past the type checker. We've already checked that metric is defined above.
+      },
+    ]
```

Comment:
This is iterating through the valid quotes, and using the first one we find. Instead, we need to do something like this:
```suggestion
return VALID_REFERENCE_RATE_QUOTES.map(quote => [quote, Number(message[`ReferenceRate${quote}`])])
                                      .filter(([quote, price]) => price >= 0)
                                      .map(([quote, price]) => ({
                                        params: {
                                          base: message.asset,
                                          quote,
                                        },
                                        value: price
                                      }))
```